### PR TITLE
Implementation for the wasi_clock_time/res

### PR DESCRIPTION
**Pull request by fadumina1**

### CODE REVIEW

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:61]: This is an undocumented, internal API so it might be worth loading it at runtime (so we don't end up calling it if it's unavailable)? Also I'm not sure if it would be available on UWP.

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:26]: I think you'll need to be careful here about integer overflow - probably add `ull` suffix to every number in the calculation?

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:49]: I'm assuming the frequency won't change for the lifetime of the app - we could calculate it once and store the result in a static variable so we don't make unnecessary system calls.

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:133]: nit: [typo] `kernel_time`

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (loganek) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:177]: worth adding a TODO here and a brief explanation - I presume we only keep it here so poll_oneoff compiles. Because we plan to refactor poll_oneoff, we'll need to remove it as part of that change.

File: core/shared/platform/include/platform_common.h
Comment (loganek) [inline core/shared/platform/include/platform_common.h:129]:
```suggestion
/**
 * Get a resolution of the clock
 *
 * @param clock_id clock identifier
 * @param resolution output variable to store the clock resolution
 * @return BHT_OK if success; otherwise, BHT_ERROR
 */
```
please do the same for `os_clock_time_get`

File: core/shared/platform/windows/win_clock.c
Comment (loganek) [inline core/shared/platform/windows/win_clock.c:38]: I think we should not ignore the return value here; errors should be propagated all the way up to the user.

File: core/shared/utils/bh_time.c
Comment (loganek) [inline core/shared/utils/bh_time.c:9]: it'd be good to mention what we convert it to, e.g.
```suggestion
convert_timespec_to_nanoseconds(const struct timespec *ts)
```

File: product-mini/platforms/windows/CMakeLists.txt
Comment (loganek) [inline product-mini/platforms/windows/CMakeLists.txt:164]: Could we merge those two into one?
```suggestion
if (WIN32)
  target_link_libraries(libiwasm ntdll)
  target_link_libraries(iwasm ntdll)
endif()
```

File: product-mini/platforms/windows/CMakeLists.txt
Comment (loganek) [inline product-mini/platforms/windows/CMakeLists.txt:166]: Unnecessary second empty line
```suggestion
```

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [reply core/shared/platform/windows/win_clock.c:61]: nvm, from what I can tell ntdll is always loaded by the process and even though this is undocumented it seems to be used quite ubiquitously anyway.

File: core/shared/platform/include/platform_api_vmcore.h
Comment (loganek) [inline core/shared/platform/include/platform_api_vmcore.h:71]: we already have the declaration in platform_common.h so the declaration should probably be removed from this file.

File: core/shared/platform/include/platform_common.h
Comment (loganek) [inline core/shared/platform/include/platform_common.h:45]: Feels like the enum fits better in the `bh_time.h` (or `bh_clock.h`) file.

File: core/shared/platform/include/platform_common.h
Comment (loganek) [inline core/shared/platform/include/platform_common.h:136]: I'm actually wondering if we should move those two declarations to `platform_api_extension.h` file.

File: core/shared/platform/windows/win_clock.c
Comment (loganek) [inline core/shared/platform/windows/win_clock.c:13]:
```suggestion
calculate_monotonic_clock_frequency(uint64 *out_frequency)
```
please use sneak case.

File: core/shared/platform/windows/win_clock.c
Comment (loganek) [inline core/shared/platform/windows/win_clock.c:40]:
```suggestion
get_performance_counter_value(uint64 *out_counter)
```

File: core/shared/platform/windows/win_clock.c
Comment (loganek) [inline core/shared/platform/windows/win_clock.c:39]:
```suggestion
int
```
Should that be `int` instead? We return `BHT_ERROR`/`BHT_OK` here.

File: core/shared/platform/windows/win_clock.c
Comment (loganek) [inline core/shared/platform/windows/win_clock.c:97]:
```suggestion
            FILETIME sys_now;
```
please follow sneak_case

File: product-mini/platforms/windows/CMakeLists.txt
Comment (loganek) [inline product-mini/platforms/windows/CMakeLists.txt:140]:
```suggestion
target_link_libraries (iwasm vmlib ${LLVM_AVAILABLE_LIBS} ${UV_A_LIBS})
```
whitespace change

**Review by loganek — CHANGES_REQUESTED** [review]

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (zoraaver) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:320]: Sorry for the late comment, did we consider implementing these functions (wasmtime_ssp_clock_time_get and wasmtime_ssp_clock_res_get) directly in each platform? Unlike the filesystem or socket functions, they don't  contain any logic to lookup a fd object from the fd table so it seems they are very small wrapper around the platform implementation anyway.

Also, since `os_clock_time_get` and `os_clock_res_get` are only used in the WASI libc implementation, we could probably define the signature for these functions in terms of WASI types (`__wasi_clockid_t` etc.) to avoid some boilerplate. What do you think?

Comment (loganek) [reply core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:320]: I think that's not been done mainly due to the circular dependency as discussed in  https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1363526897 

I think we can merge this as it is now, but once your change @zoraaver is merged, we'll make an update to this code and simplify it if possible. How does it sound to you?

**Comment by loganek** [comment]
@wenyongh would you be able to have a look at this PR?

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (zoraaver) [reply core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:320]: Agree, I think the PR is in a good state to merge now and we can simplify the code later once #2585  is merged.

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:182]: How about renaming to wasi_clockid_to_clockid?

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:287]: How about renaming to wasi_clockid_to_bh_clockid?

File: core/shared/platform/common/posix/posix_clock.c
Comment (wenyongh) [inline core/shared/platform/common/posix/posix_clock.c:10]: How about renaming to bh_clockid_to_clockid? And found that there is a different implementation in `core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c` with different function prototype?

File: core/shared/utils/bh_time.c
Comment (wenyongh) [inline core/shared/utils/bh_time.c:9]: How about renaming to bh_timespec_to_nanoseconds? Adding bh_ prefix is because that APIs in core/shared/utils usually start with it.

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:177]: Should be `placeholder`

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (wenyongh) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:183]: Invalid coding style, could you use clang-format-12 to format the code?
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/6573605221/job/17856981299

File: core/shared/platform/windows/win_clock.c
Comment (wenyongh) [inline core/shared/platform/windows/win_clock.c:50]: Had better remove unused line?

File: core/shared/platform/windows/win_clock.c
Comment (wenyongh) [inline core/shared/platform/windows/win_clock.c:132]: Had better put this line before L131

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (loganek) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:180]: I guess this comment is no longer relevant?

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:129]: The braces for each case in the switch statement aren't strictly necessary but it looks a bit weird to have the return statement outside the block - maybe move it a line above so that it's inside the block?

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:60]: In this file, there are blank lines after some of the if statements and before some of the 'case' blocks in the switch statement. I think we can remove them.

File: core/shared/platform/windows/win_clock.c
Comment (zoraaver) [inline core/shared/platform/windows/win_clock.c:25]: typo 'derives'

File: core/shared/platform/common/posix/posix_clock.c
Comment (zoraaver) [inline core/shared/platform/common/posix/posix_clock.c:7]: We can probably remove this header include now

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (zoraaver) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:98]: We can keep the static here I think

File: core/shared/platform/common/posix/posix_clock.c
Comment (zoraaver) [reply core/shared/platform/common/posix/posix_clock.c:40]: Or define a macro `#define NANOSECONDS_PER_SECOND 1000000000ULL` like in `win_clock.c`

File: core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
Comment (zoraaver) [inline core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:200]: we can make this static

File: core/shared/platform/common/posix/posix_clock.c
Comment (wenyongh) [inline core/shared/platform/common/posix/posix_clock.c:9]: Had better make this function static?

File: core/shared/platform/windows/win_clock.c
Comment (wenyongh) [inline core/shared/platform/windows/win_clock.c:28]: How about renaming to `file_time_to_wasi_timestamp` like others?

File: core/shared/platform/windows/win_clock.c
Comment (wenyongh) [inline core/shared/platform/windows/win_clock.c:39]: Had better make this function static?

**Review by zoraaver — APPROVED** [review]
LGTM

**Review by wenyongh — COMMENTED** [review]
LGTM

**Review by loganek — APPROVED** [review]
Good job!

**Pull request merged**
